### PR TITLE
Implement support to abort a running execution

### PR DESCRIPTION
**Pull request by dmitrys-odysseus**

### CODE REVIEW

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    private ExecutionOutcome abort(Long id, ROverseer overseer) {
+        CompletableFuture<ExecutionOutcome> future = overseer.abort();
+        try {
+            return future.get(30, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            log.info("Execution [{}] waiting for abort operation was interrupted", id);
+            return new ExecutionOutcome(Stage.ABORT, "Waiting for abort interrupted", overseer.getStdout());
+        } catch (ExecutionException e) {
+            log.info("Execution [{}] abort attempt failed", id, e);
+            overseer.getStdout();
```
Comment (wivern) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java:141]: What does this line?

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            return new ExecutionOutcome(Stage.ABORT, e.getMessage(), overseer.getStdout() + "\n" + getStackTrace(e.getCause()));
+        }
+    }
+
+    private void attachMetadata(AnalysisSyncRequestDTO analysis, File analysisDir) {
+        DataSourceUnsecuredDTO dataSource = analysis.getDataSource();
+        try {
+            cdmMetadataService.extractMetadata(dataSource, analysisDir);
+        } catch (Exception e) {
+            log.info("Execution [{}] error collecting CDM metadata for datasource [{}]", analysis.getId(), dataSource.getName(), e);
```
Comment (wivern) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java:154]: The most often case when metadata extraction is failed, is an empty Metadata table. I'd like to suggest not to print full stack trace into the info channel. If it seems useful, then put stacktrace into debug.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/Overseer.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package com.odysseusinc.arachne.executionengine.execution;
+
+import com.odysseusinc.arachne.execution_engine_common.api.v1.dto.ExecutionOutcome;
+import java.time.Instant;
+import java.util.concurrent.CompletableFuture;
+
+public interface Overseer {
```
Comment (wivern) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/Overseer.java:7]: Supervisor might be more evident

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    public ROverseer(long id, Process process, int timeout, BiConsumer<String, String> callback, int updateInterval, Instant started) {
+        this.id = id;
+        this.process = process;
+        reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+        this.callback = callback;
+        this.started = started;
+
+        result = new CompletableFuture<>();
+
+        ScheduledFuture<?> watchdog = executor.schedule(() -> {
```
Comment (wivern) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java:55]: The watchdog variable is unused. Moreover, what this scheduled task is doing?

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        }
+        return dead;
+    }
+
+    private boolean waitForKill() {
+        try {
+            // Will not kill whole process tree on windows. The fundamental problem here is that (unlike Unix)
+            // Windows doesn't maintain parent-child relationships between processes. A process can kill
+            // its own immediate children, but not 'grand-children' because it has no way of finding them.
+            return process.destroyForcibly().waitFor(30, TimeUnit.SECONDS);
```
Comment (wivern) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java:145]: Probably it's better to make timeout configurable

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java (hunk 2 above)
Comment (dmitrys-odysseus) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java:154]: As far as I tested, on an empty metadata table, analysis will not even start.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java (hunk 1 above)
Comment (dmitrys-odysseus) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java:141]: Nothing. Removed

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java (hunk 4 above)
Comment (dmitrys-odysseus) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java:55]: The variable is indeed unnecessary and is removed.
The scheduled task will terminate the execution after `runtime.timeOutSec` seconds.
This is not a new behaviour, however with this PR it is improved to end up with a real termination.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java (hunk 5 above)
Comment (dmitrys-odysseus) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java:145]: Good catch. I was planning to do so initially but overlooked it later on.
Added optional setting: `runtime.killTimeoutSec`

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java (hunk 2 above)
Comment (wivern) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/AnalysisService.java:153]: Out of metadata should not impact analysis. Metadata gathering might be a problem, that's correct. But I'd like to separate analysis issues from metadata issues, and not to see a dozen of useless stacktrace on each analysis submission. Analysis should start regardless of metadata.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java (hunk 4 above)
Comment (wivern) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java]: Well, then 1 hour is too few. There are analyses that can run about 10-12 hours.

**Review by wivern — APPROVED** [review]

**Pull request merged**